test/dashboardHandler.test.js

```diff
@@ -0,0 +1,369 @@
+      dashboard: "./sample-data/dashboard-multiple-sample.json",
+      uniqueServices: 51,
+      expectTruncated: true,
+
+      metrics: {
+        "Metrics": [].concat.apply([], generateMetrics(51))
+      },
+      event: {
+        "account": "123456789012",
+        "region": "ap-southeast-2",
+        "detail": {},
+        "detail-type": "Scheduled Event",
+        "source": "aws.events",
+        "time": "2019-03-01T01:23:45Z",
+        "id": "cdc73f9d-aea9-11e3-9d5a-835b769c0d9c",
+        "resources": [
+          "arn:aws:events:ap-southeast-2:123456789012:rule/my-schedule"
+        ]
+      }
+    }
+  ]
+
+  scenarios.forEach(scenario => {
+    describe(scenario.description, () => {
+      beforeEach(() => {
+
+        // https://github.com/dwyl/aws-sdk-mock/issues/118
+        // Cannot use aws-sdk-mock
+        putDashboardSpy = sandbox.stub().returns({
+          // putDashboardSpy = sinon.stub().returns({
+          promise: () => Promise.resolve()
+        })
+
+        listMetricsStub = sandbox.stub().returns({
+          // listMetricsStub = sinon.stub().returns({
+          eachPage: (cb) => {
+            cb(null, scenario.metrics)
+
+            cb(null, null) // no more pages
+          }
+        })
+        // sinon.stub(awsSdk, 'CloudWatch').returns({
+        sandbox.stub(awsSdk, 'CloudWatch').returns({
+          listMetrics: listMetricsStub,
+          putDashboard: putDashboardSpy
+        });
+
+        awsSdk.config.region = 'ap-southeast-2';
+      })
+
+      afterEach(() => {
+        // awsSdk.CloudWatch.restore();
+        sandbox.restore();
+      })
+
+      it("should generate a dashboard", () => {
+        return LambdaTester(index.generateDashboardTrend)
+          .event(scenario.event)
+          .expectResult((result, additional) => {
+            expect(putDashboardSpy).to.have.callCount(1);
+          });
+      })
+
+      it('should generate 3 text widgets - to explain each metric + interpretation', () => {
+        return LambdaTester(index.generateDashboardTrend)
+          .event(scenario.event)
+          .expectResult((result, additional) => {
+            const dashboard = JSON.parse(putDashboardSpy.getCall(0).args[0].DashboardBody);
+            const textWidgets = dashboard.widgets.filter(w => w.type === 'text');
+
+            expect(textWidgets.length).to.equal(3);
+
+          });
+      })
+
+      if (scenario.expectTruncated) {
+        describe('When there are too many metrics in the account', () => {
+          it('should report a maximum of 150 metrics in the dashboard', () => {
+            const consoleSpy = sandbox.spy(console, 'warn')
+            return LambdaTester(index.generateDashboardTrend)
+              .event(scenario.event)
+              .expectResult((result, additional) => {
+                expect(consoleSpy).to.have.been.calledWith("Maximum of 150 metrics are allowed in a single dashboard. Some metrics will not be reported.");
```

Comment:
This is shown in CloudWatch only - the requirement is to report it in the dashboard somewhere.

Reply:
done